File: Stacks & Queues/Queue.py
# QUEUES WITH LINKED LIST

class Node:
    def __init__ (self, value):
        self.value = value
        self.next = None
    

class LinkedQueue:
    def __init__ (self):
        self.first = None
        self.last = None
        self.length = 0

    def isEmpty(self):
        if self.first == None:
            return True
        else: return False

    def peek(self):
        return self.last.value
    
    def enqueue(self, value):
        newNode = Node(value)
        if self.isEmpty() == True:
            self.first = newNode
            self.last = newNode
        elif self.first == self.last:
            self.first.next == newNode
            self.last = newNode
        else:
            self.last.next == newNode
            self.last = newNode
        self.length +=1
        return self.last.value

    def dequeue(self):
        if not self.first:
            return "Queue is Empty"
        elif self.first == self.last:
            self.first = None
            self.last = None
            return "Queue is Empty"
        else:
            # firstNode = self.first
            self.first = self.first.next
            # firstNode = None
        self.length -=1
        return self.first



# TESTING

# myQueue = LinkedQueue()
# print(myQueue.isEmpty())
# print(myQueue.peek())
# print(myQueue.enqueue("Lovelle"))
# print(myQueue.enqueue("Jade"))
# print(myQueue.enqueue("Saphir"))
# print(myQueue.dequeue())
# print(myQueue.peek()) 

File: Stacks & Queues/Queue.js
// QUEUES USING LINKED LIST

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
} 

class LinkedQueue {
    constructor() {
        this.first = null;
        this.last  = null;
        this.length = 0;
    }
    isEmpty(){
        if (this.first === null){
            return true
        }else {
            return false
        }

    }
    peek() {
        if (this.isEmpty() === true){
            return "Queue is Empty"
        }else{
            return this.first;
        }
    }
    enqueue(value) {
        const newNode = new Node(value);
        newNode.value = value
        if (this.length === 0) {
            this.first = newNode;
            this.last = newNode;
        }else {
            this.last.next = newNode
            this.last = newNode
        }
        this.length++;
        return this
    }
    dequeue() {
        if (!this.first){
            return ("Queue is empty")
        }
        if (this.first == this.last) {
            this.last = null 
        }
        this.first = this.first.next 
        this.length--;
        return this
    }
}

// TESTING 
const myQueue = new LinkedQueue()
console.log(myQueue.peek())
console.log(myQueue.enqueue("Joy"))
console.log(myQueue.enqueue("Kelvin"))
console.log(myQueue.enqueue("Collins"))
console.log(myQueue.enqueue("Carlos"))
console.log(myQueue.enqueue("Logan"))
console.log(myQueue.dequeue());
console.log(myQueue.dequeue());
console.log(myQueue.dequeue());


// QUEUE USING A STACK

